### ISSUE TITLE
Percentage oddity

### ISSUE THREAD

**Issue by sergeevabc** [PaddiM8/kalker#53]
``` console
$ kalker.exe -10.5%+10.5%
-0.116025
```

Great, but it does not make sense. [Qalc][1], for example, outputs 0 in this case.

[1]: https://qalculate.github.io/

**Comment by wischi-chr** [comment]
It's not really odd, it depends on how you interpret the input.

Percent is just a fancy way of writing scaled factors. So 50% just means 0.5, 10% is 0.1, etc.
One way of interpreting your calculation would be to expand the percent values to -0.105 + 0.105 which would be 0 (I guess that was how Qalc interpreted the calculation)

But that's not the only way a calculator can interpret the percent sign. For convenience calculators support "adding/subtracting" percentages which are technically multiplications and divisions.

Let's take for example `100 + 10%` if we interpret it like before that would be 100 + 0.1 = 100.1. But that is clearly not the result the user expected. So "adding percentages" are really multiplications and `100 + 10%` will be interpreted as `100 * 1.1` which is 110. You can see this in action if you change the order in Qcal. Addition in general is commutative but "percentage addition" is not because the "+ 10%" is multiplied to the value that came before ... if there is no value before you don't have a base value and it will be reinterpreted just as a number/100 so 10% will become 0.1.

Qcal:
```
100 + 10% -> 110
10% + 100 -> 100.1
```

this is also what happens in kalker with your calculation `-10.5%` cannot be applied to anything that stands before it because it's the first expression, so it's evaluated to `0.105` now the next part `+10.5%` is applied to the value left of it. Remember "percentage add" is a multiplication so `0.105+10.5%` is evaluated (also in Qcal) as `0.105*1.105` which is `0.116025`

So neither Qalc nor kalker are _"right"_ or _"wrong"_ here, it's basically an edge case because percentages can be interpreted in two different ways and it's not always clear what the user meant.

My tip: only do very basic stuff if you use the percentage operator. If you need to do advanced stuff be more explicit and don't use the percent sign in you calculations.

**Comment by aukeroorda** [comment]
Just leaving my thoughts - the approach of using a percentage suffix to write fractions is okay-ish, but having their value based on the preceding numeric value is such a strange approach. Of course it matches with the natural language description of these calculations, but the conversion to a proper computation of the value should be done by the users of this program, and not as an ambiguous (modulo or percentage operation?) implicit conversation step in this program, especially if you target a scientific audience. It feels out of place in this otherwise well designed program.

**Comment by timokoesters** [comment]
I agree, percentage for me is just a shorthand for `1/100`. But maybe there should be a flag to toggle the behavior, if it is a common convention for others.

**Comment by PaddiM8** [comment]
It's quite common for calculators to base the result on the preceding value (WolframAlpha does it too) and I use it quite a lot myself, to do things like `130 - 25%`, but yes it can lead to confusion. Having a flag to disable it would probably be reasonable.

**Comment by NobbZ** [comment]
The way `kalker` and many other similar tools handle percentage is (according to 10 years of school) mathematically incorrect.

And even though this shorthand makes sense for a regular pocket calculator, it doesn't really make sense for a REPL style calculator.

Due to the perceived misbehaviour in percentages, I can not build trust in the remaining system to be correct.

Due to this misbehaviour in percentages, simple properties of addition do not hold anymore, as `1 + 5% != 5% + 1` and `(1 + 5%) + 5% != 1 + (5% + 5%)`.

And even worse, during experiments I learned, that `5% + 5%` is `5.25%`, which makes even less sense…

`libqalc` does seem to get it correct to my intuition.

**Comment by tejing1** [comment]
I second the idea of an option to make `%` just a shorthand for `/100`, instead of this weirdness.

**Comment by wischi-chr** [comment]
@tejing1 

One could certainly do that, but that has probably a lot of unexpected results for many people.

For example what would you expect the result of `100 + 10%` to be? With your interpretation that `%` just being a shorthand for `/` the result would be `100.1` and probably not what most people would expect.

I'm not a fan of the percent operator at all. Not specifically in kalker, but in general, because the implementation is always different and often with magic edge cases, so one would typically first have to test how the percent operator behaves before using it, so I don't use it at all, because it's way simpler to just use multiplication.

**Comment by tejing1** [comment]
I would expect `100 + 10%` to be `100.1`, yes. I find the other behavior baffling in practice. It follows an abuse of notation that has never had well-defined semantics, and those frayed edges of the semantics really show up when you try to systematically encode the notation into a program.

In the process of investigating this a little more, I also learned that `a/b*c` is parsed as `a/(b*c)` in kalker as well... Which I have never seen in another expression parser in my life before. This is bringing me around more to @NobbZ' perspective. kalker isn't the calculator for me. It has some nice aspects, but I can't depend on it to understand what I want it to do, which is critical to it fulfilling its role as a calculator at all, and I'm not going to upend my notations of expression syntax for its sake alone.

**Comment by wischi-chr** [comment]
> In the process of investigating this a little more, I also learned that a/b*c is parsed as a/(b*c) in kalker as well...

I'm pretty sure you are wrong on that one:

![image](https://github.com/user-attachments/assets/ae995baa-14f9-4534-8f88-5680e7091b04)

Regarding the percentage thing. The percentage operator should never have been added to any calculator if you ask me, because it's very hard to add it in a meaningful and consistent way, so IMHO the percentage sign should be dropped altogether. I don't see a benefit in having a shorthand for `/100`

**Comment by PaddiM8** [comment]
The percentage operator is simply there as a convenience thing. I agree that it isn't a "good" operator because of the inherent inconsistencies, but I don't see why you it would be useful for it to be a shorthand for /100, which means that it doesn't really hurt to have it, in my opinion. A lot of people are used to using this operator for convenience when doing quick calculations. If you, for understandable reasons, don't like it, then you can just ignore it.

**Comment by tejing1** [comment]
@wischi-chr I'm sorry, you're correct. I misinterpreted what went wrong:
```
>> 4/√4*3
≈ 1.1547005384
>> (4/√4)*3
= 6
>> 4/(√4)*3
= 6
>> 4/√(4*3)
≈ 1.1547005384
>> 
```
My issue was caused by the `√` being lower precedence than `*`, rather than anything to do with the `/`. Which is also surprising, but less egregious, I suppose?
